lib/srv/sess_test.go

```diff
@@ -988,3 +993,176 @@ func TestSessionRecordingMode(t *testing.T) {
+// user.
+func TestCloseRemoteSession(t *testing.T) {
+	srv := newMockServer(t)
+	srv.component = teleport.ComponentProxy
+
+	// init a session registry
+	reg, _ := NewSessionRegistry(SessionRegistryConfig{
+		Srv:                   srv,
+		SessionTrackerService: srv.auth,
+	})
+	t.Cleanup(func() { reg.Close() })
+
+	scx := newTestServerContext(t, reg.Srv, nil)
+	scx.SessionRecordingConfig.SetMode(types.RecordAtProxy)
+	scx.RemoteSession = mockSSHSession(t)
+
+	sid := string(rsession.NewID())
+	scx.SetEnv(sshutils.SessionEnvVar, sid)
+
+	// Open a new session
+	sshChanOpen := newMockSSHChannel()
+	// Always close the session from the client side to avoid it being stuck
+	// on closing (server side).
+	t.Cleanup(func() { sshChanOpen.Close() })
+	go func() {
+		// Consume stdout sent to the channel
+		io.ReadAll(sshChanOpen)
+	}()
+
+	err := reg.OpenSession(context.Background(), sshChanOpen, scx)
+	require.NoError(t, err)
+	require.NotNil(t, scx.session)
+
+	// After the session is open, we force a close comming from the server. Do
+	// this inside a goroutine to avoid being blocked.
+	closeChan := make(chan error)
+	go func() {
+		closeChan <- scx.session.Close()
+	}()
+
+	select {
+	case err := <-closeChan:
+		require.NoError(t, err)
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "expected session to be closed")
+	}
+}
+
+func mockSSHSession(t *testing.T) *tracessh.Session {
+	ctx := context.Background()
+
+	_, key, err := ed25519.GenerateKey(nil)
+	require.NoError(t, err)
+	signer, err := ssh.NewSignerFromKey(key)
+	require.NoError(t, err)
+
+	cfg := &ssh.ServerConfig{NoClientAuth: true}
+	cfg.AddHostKey(signer)
+
+	listener, err := net.Listen("tcp", "localhost:")
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+
+	go func() {
+		conn, err := listener.Accept()
+		if err != nil {
+			t.Logf("error while accepting ssh connections: %s", err)
+		}
+
+		srvConn, chCh, reqCh, err := ssh.NewServerConn(conn, cfg)
+		go ssh.DiscardRequests(reqCh)
+		go func() {
+			for newChannel := range chCh {
+				channel, requests, err := newChannel.Accept()
+				if err != nil {
+					t.Logf("Failed to handle new channel: %s", err)
+					continue
+				}
+
+				go func() {
+					for req := range requests {
+						req.Reply(true, nil)
+					}
+				}()
+
+				sessTerm := term.NewTerminal(channel, "> ")
+				go func() {
+					defer channel.Close()
+					for {
+						line, err := sessTerm.ReadLine()
+						if err != nil {
+							break
+						}
+						t.Log(line)
+					}
+				}()
+			}
+		}()
+
+		srvConn.Wait()
+	}()
+
+	// Establish a connection to the newly created server.
+	sessCh := make(chan *tracessh.Session)
+	go func() {
+		client, err := tracessh.Dial(ctx, listener.Addr().Network(), listener.Addr().String(), &ssh.ClientConfig{
+			User:            "user",
+			Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
+			HostKeyCallback: ssh.FixedHostKey(signer.PublicKey()),
+		})
+		require.NoError(t, err)
```

Comment:
`require` in this specific goroutine works but the outer goroutine is left waiting for a timeout. I don't see a sensible way to use `ssh.Dial` (like, ever), so maybe the solution is to split the tcp dialing (with a timeout, and a timer that kills the connection after a while) and the SSH protocol stuff, and just do everything in a single goroutine?

```go
timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
defer cancel()

nc, err := new(net.Dialer).DialContext(timeoutCtx, listener.Addr().Network(), listener.Addr().String())
require.NoError(t, err)
defer nc.Close()
stop := context.AfterFunc(timeoutCtx, func() { _ = nc.Close() })
defer stop()

huey, dewey, louie, err := ssh.NewClientConn(nc, listener.Addr().String(), &ssh.ClientConfig{...})
require.NoError(t, err)
client := ssh.NewClient(huey, dewey, louie)

sess, err := client.NewSession()
require.NoError(t, err)
require.NoError(t, sess.Close())
```

Reply:
The timeout is not applied on the SSH handshake, which is the flaw that makes `ssh.Dial` basically unusable in real production code.